src/Roave/SecurityAdvisories/Constraint.hs

```diff
@@ -0,0 +1,140 @@
+module Roave.SecurityAdvisories.Constraint
+  ( Version(..)
+  , VersionLimit(..)
+  , VersionBoundary(..)
+  , VersionRange(..)
+  , VersionConstraintUnion(..)
+  , makeVersion
+  , canMergeRanges
+  , stringToBoundary
+  , stringToVersionLimit
+  , stringToVersion
+  ) where
+
+import Data.List
+import Data.List.NonEmpty
+import qualified Data.List.NonEmpty as N
+import Numeric.Natural
+
+import Data.Void
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer as L
+import Text.Megaparsec.Expr
+
+data Version =
+  Version (NonEmpty Natural)
+  deriving (Eq, Ord)
+
+data VersionLimit
+  = LessThanEquals
+  | LessThan
+  | Equals
+  | GreaterThan
+  | GreaterThanEquals
+  deriving (Eq, Show)
+
+data VersionBoundary =
+  VersionBoundary VersionLimit
+                  Version
+  deriving (Eq, Show)
+
+data VersionRange
+  = Range VersionBoundary
+          VersionBoundary
+  | From VersionBoundary
+  | Till VersionBoundary
+  deriving (Eq, Show)
+
+data VersionConstraintUnion =
+  ConstraintUnion (NonEmpty VersionRange)
+  deriving (Eq, Show)
+
+instance Show Version where
+  show (Version v) = intercalate "." (toList (N.map show v))
```

Comment:
If you declare Version like this:

```haskell
newtype Version = Version
  { getVersion :: NonEmpty Natural
  } deriving (Eq, Ord)
```

you can write the Show instance for Version like this:

```haskell
instance Show Version where
  show = intercalate "." . toList . N.map show . getVersion
```

which is of course superior to any other way of writing it because it's point-free! </sarcasm>
Tip: Read (.) as "after". "Intercalate after converting it to a list after mapping show".

Reply:
A note regarding the Show instance and the abuse thereof: http://hackage.haskell.org/package/base-4.11.1.0/docs/Text-Show.html#t:Show

It's intended to be used with Read and should follow very strict semantics. Everyone is kinda abusing it to just convert data into a string however. Just something to keep in mind.

Reply:
I don't think I would change `Version` to have `getVersion` be the field - probably rather define `getVersion :: Version -> NonEmpty Natural` as an independent function

Reply:
It's a common pattern to access `newtype` contents. Mostly named `newtype Foo = Foo { getFoo :: X }` or `newtype Foo = Foo { unFoo :: X }`. It behaves the same as a function defined elsewhere, but is easier to declare.

Reply:
Yeah, but declaration is not really that interesting anyway